python/vast/utils/arrow.py

```diff
@@ -1,7 +1,125 @@
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return self.ext_name.encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        if serialized.decode() != self.ext_name:
+            raise TypeError("type identifier does not match")
+        if storage_type != self.ext_type:
+            raise TypeError("storage type does not match")
+
+    def __reduce__(self):
+        return AddressScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return AddressScalar
+
+
+class SubnetScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Network | ip.IPv6Network:
+        address = unpack_ip(self.value[0].as_py())
+        length = self.value[1].as_py()
+        return ip.ip_network((address, length), strict=False)
+
+
+class SubnetType(pa.ExtensionType):
+    ext_name = "vast.subnet_type"
+    # This gives us a pyarrow.lib.ArrowNotImplementedError :-/
+    # ext_type = pa.struct([("address", AddressType()),
+    #                       ("length", pa.uint8())])
+    ext_type = pa.struct([("address", pa.binary(16)), ("length", pa.uint8())])
+
+    def __init__(self):
+        pa.ExtensionType.__init__(self, self.ext_type, self.ext_name)
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return self.ext_name.encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        if serialized.decode() != self.ext_name:
+            raise TypeError("type identifier does not match")
+        if storage_type != self.ext_type:
+            raise TypeError("storage type does not match")
+        return SubnetType()
+
+    def __reduce__(self):
+        return SubnetScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return SubnetScalar
+
+
+class EnumScalar(pa.ExtensionScalar):
+    def as_py(self) -> str:
+        return self.type.field(self.value.as_py())
+
+
+class EnumType(pa.ExtensionType):
+    ext_name = "vast.enumeration_type"
+    ext_type = pa.uint32()
+
+    def __init__(self, fields: dict[int, str]):
+        self._fields = fields
+        pa.ExtensionType.__init__(self, self.ext_type, self.ext_name)
+
+    @property
+    def fields(self):
+        return self._fields
+
+    def field(self, key):
+        return self._fields[key]
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return json.dumps(self._fields).encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        fields = json.loads(serialized.decode())
+        if storage_type != self.ext_type:
+            raise TypeError("storage type does not match")
+        return EnumType(fields)
+
+    def __reduce__(self):
+        return EnumScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return EnumScalar
+
+
+# TODO: move to appropriate location
+pa.register_extension_type(AddressType())
+pa.register_extension_type(SubnetType())
```

Comment:
```suggestion
pa.register_extension_type(AddressType())
pa.register_extension_type(SubnetType())
pa.register_extension_type(EnumType())
```

Reply:
We can't register an `EnumType` without fields.

Since we don't know all possible enums in advance, I wonder how this chicken-and-egg issue will work with IPC.

Reply:
Now that I added the tests for IPC, I have a data point: Arrow needs the concrete extension type to be registered, otherwise the roundtrip fails. How do we get around this in C++ land?

Reply:
You just need to register one instance of the extension type:

https://github.com/tenzir/vast/blob/1c3eec37e96db254ad73a405164718b4d435d779/libvast/src/type.cpp#L2138-L2144

Reply:
Got it. I added the same stub enum type registration.